vignettes/converting_files.Rmd

```diff
@@ -0,0 +1,97 @@
+
+## Converting from XPT
+
+Ideally if you're converting from an XPT file or a SAS7BDAT file, the appropriate metadata would be maintained separate from the dataset itself. Inevitably some information will get lost while being exchanged because of differences of what information was applied or how R interprets SAS datasets. This is one reason why **{datasetjson}** improves upon XPTs as an exchange format. In leiu of external metadata, here's an example of how you can make a best effort conversion.
+
+```{r XPT conversion}
+adsl <- haven::read_xpt(file.path(system.file(package='datasetjson'), "adsl.xpt"))
+
+#' Gather variable metadata in Dataset JSON compliant format
+#'
+#' @param n Variable name
+#' @param .data Dataset to gather attributes
+#'
+#' @returns Columns compliant data frame
+extract_xpt_meta <- function(n, .data) {
+  
+  attrs <- attributes(.data[[n]])
+  
+  out <- list()
+
+  # Identify the variable type
+  if (inherits(.data[[n]],"Date")) {
+    out$dataType <- "date"
+    out$targetDataType <- "integer"
+  } else if (inherits(.data[[n]],"POSIXt")) {
+    out$dataType <- "datetime"
+    out$targetDataType <- "integer"
+  } else if (inherits(.data[[n]],"numeric")) {
+    if (any(is.double(.data[[n]]))) out$dataType <- "float"
+    else out$dataType <- "integer"
+  } else {
+    out$dataType <- "string"
+    out$length <- max(purrr::map_int(.data[[n]], nchar))
+  }
+  
+  out$itemOID <- n
+  out$name <- n
+  out$label <- attr(.data[[n]], 'label')
+  out$displayFormat <- attr(.data[[n]], 'format.sas')
+  tibble::as_tibble(out)
+  
+}
+
+# Loop the ADSL columns
+adsl_meta <- purrr::map_df(names(adsl), extract_xpt_meta, .data=adsl)
+
+# Create the datasetjson object
+ds_json <- dataset_json(
+  adsl, 
+  item_oid = "ADSL",
+  name = "ADSL",
+  dataset_label = attr(adsl, 'label'),
+  columns = adsl_meta
+)
+
+# Write the JSON
+json_file_content <- write_dataset_json(ds_json)
+
+# Check schema compliance
+validate_dataset_json(json_file_content)
+
+```
+
+## Bulk File Conversion
+
+If your intention is to convert files into Dataset JSON format in bulk, there are a couple things you should consider - especially if you're trying to replicate existing procedures done using SAS:
+
+Remember that R by default holds data in memory, whereas work datasets in SAS are still written to disk. This means that if you're doing a bulk conversion of datasets, you'll want to read in and write out one dataset at a time. For example:
+
+1. Read in the XPT or SAS7BDAT file
+2. Write out the Dataset JSON file
+3. Remove the objects from memory to free up space
+
+It's likely best to wrap the conversion process in a function, because the function namespace will inherently release the temporary objects during garbage collection. Depending on the size of your data, it could be easy to max out memory if you first read in all the datasets.
```

Comment:
Should we provide an example function. I need to make this for internal testing anyways.  I could include a "somthing like this ...." as a reference.

Reply:
Yeah that's fine - go for it.

Reply:
I'll add this once I have it read.  It's not needed right this second.